### PR TITLE
install-grub.pl: Do stat() last to not hang on hanging network FS

### PR DESCRIPTION
**Pull request by nh2**
## Description of changes

Fixes `install-grub.pl` (and thus `nixos-rebuild`) hanging indefinitely when there are hanging network file systems.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

I've tested this on our production servers over the last 6 months.

### CODE REVIEW

**Comment by nh2** [comment]
CCing some other recent `install-grub.pl` authors as potential reviewers:

@SrTobi @arianvp @mkg20001 @dmadisetti @thiagokokada @volth @vmandela

File: nixos/modules/system/boot/loader/grub/install-grub.pl [hunk 1]
```diff
@@ -155,6 +154,11 @@ sub GetFs {
 
         # Is it better than our current match?
         if (length($mountPoint) > length($bestFs->mount)) {
+
+            # Check if it's a real directory. Do this last, because it does a stat(),
```
Comment (dmadisetti) [inline nixos/modules/system/boot/loader/grub/install-grub.pl:158]: Is this the only bottleneck? Looks like there are a few mktmp / tree walking operations throughout the script

but LGTM as an inconsequential change.

**Review by dmadisetti — APPROVED** [review]

File: nixos/modules/system/boot/loader/grub/install-grub.pl (hunk 1 above)
Comment (nh2) [reply nixos/modules/system/boot/loader/grub/install-grub.pl:158]: > Is this the only bottleneck? Looks like there are a few mktmp / tree walking operations throughout the script

As far as I could tell, yes.

The tree walking operations wouldn't go into arbitrary mount points, but this stat did.

**Comment by thiagokokada** [comment]
CC @roberth.

**Review by roberth — COMMENTED** [review]
Looks acceptable to me, if you could update the comment.

It's Perl, so I could never say "looks good to me", but I guess it means the same.

**Comment by roberth** [comment]
Could you add a test? NixOS tests are well suited, as you can kill, say, an NFS host.

**Comment by nh2** [comment]
> Looks acceptable to me, if you could update the comment.

@roberth Which comment / update in which way?

> Could you add a test? NixOS tests are well suited, as you can kill, say, an NFS host.

That would be nice, but I don't think I'll be able to get to that soon (currently busy trying to fix Ceph in #281858).

File: nixos/modules/system/boot/loader/grub/install-grub.pl [hunk 2]
```diff
@@ -155,6 +154,11 @@ sub GetFs {
...
+            # which may hang forever on network file systems.
```
Comment (roberth) [inline nixos/modules/system/boot/loader/grub/install-grub.pl:159]:
```suggestion
            # -d performs a stat, which can hang forever on network file systems,
            # so we only make this call last, when it's likely that this is the mount point we need.
```

**Review by roberth — COMMENTED** [review]
> > Looks acceptable to me, if you could update the comment.
> 
> @roberth Which comment / update in which way?

GitHub ate my review :t-rex:

File: nixos/modules/system/boot/loader/grub/install-grub.pl (hunk 2 above)
Comment (roberth) [reply nixos/modules/system/boot/loader/grub/install-grub.pl:159]: ie add a bit more why to the how

**Comment by nh2** [comment]
> GitHub ate my review 🦖

Thanks, done!

Merging this now, hoping it won't break anybody!

**Pull request merged**